docs/src/tutorial/cython_tutorial.rst

```diff
@@ -127,91 +130,142 @@ Primes
+
+        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.py
+            :lines: 2,3
+            :dedent:
+            :lineno-start: 2
+
+        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.py
+            :lines: 11,12
+            :dedent:
+            :lineno-start: 11
+        
+        Lines 2, 3, 11 and 12 use the variable annotations
+        to define some local C variables.
+        The result is stored in the C array ``p`` during processing,
+        and will be copied into a Python list at the end (line 26).
 
-    cdef int n, i, len_p
-    cdef int p[1000]
+    .. group-tab:: Cython
 
-Lines 2 and 3 use the ``cdef`` statement to define some local C variables.
-The result is stored in the C array ``p`` during processing,
-and will be copied into a Python list at the end (line 22).
+        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.pyx
+            :lines: 2,3
+            :dedent:
+            :lineno-start: 2
+
+        Lines 2 and 3 use the ``cdef`` statement to define some local C variables.
+        The result is stored in the C array ``p`` during processing,
+        and will be copied into a Python list at the end (line 26).
 
 .. NOTE:: You cannot create very large arrays in this manner, because
-          they are allocated on the C function call :term:`stack<Stack allocation>`, which is a
-          rather precious and scarce resource.
+          they are allocated on the C function call :term:`stack<Stack allocation>`,
+          which is a rather precious and scarce resource.
           To request larger arrays,
           or even arrays with a length only known at runtime,
           you can learn how to make efficient use of
           :ref:`C memory allocation <memory_allocation>`,
           :ref:`Python arrays <array-array>`
           or :ref:`NumPy arrays <memoryviews>` with Cython.
 
-::
-
-    if nb_primes > 1000:
-        nb_primes = 1000
+.. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.pyx
+    :lines: 5,6
+    :dedent:
+    :lineno-start: 5
 
 As in C, declaring a static array requires knowing the size at compile time.
 We make sure the user doesn't set a value above 1000 (or we would have a
-segmentation fault, just like in C).  ::
+segmentation fault, just like in C)
+
+.. tabs::
+    .. group-tab:: Pure Python
+
+        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.py
+            :lines: 8,9
+            :dedent:
+            :lineno-start: 8
 
-    len_p = 0  # The number of elements in p
-    n = 2
-    while len_p < nb_primes:
+        If we run the code from python, then we must initialize the array,
+        the easiest way is to fill it with zeros as on lines 8-9.
+        If we compile this, then the array will already be filled with
+        random data from the allocated memory, since we will change them anyway,
+        we do not need to overwrite this random data.
```

Comment:
```suggestion
        When we run this from Python we have to initialize the list,
        which is easiest done filling it with 0s (as seen on lines 8-9).
        When we compile this with Cython on the other hand, the array
        will be a 1000 items long with random data from previous allocations
        that we will overwrite.
```

Reply:
> When we compile this with Cython on the other hand, the array
> will be a 1000 items long with random data from previous allocations
> that we will overwrite.

random -> arbitrary. The data is not random, that's a common misperception that we should not spread here.

```suggestion
        When we run this code from Python, we have to initialize the array.
        This is most easily done by filling it with zeros (as seen on line 8-9).
        When we compile this with Cython, on the other hand, the array will
        behave as in C.  It is allocated on the function call stack with a fixed
        length of 1000 items that contain arbitrary data from the last time that
        memory was used.  We will then overwrite those items in our calculation.
```

Reply:
> The data is not random, that's a common misperception that we should not spread here.

or we can say pseudo-random to be true.

Reply:
No Scoder is right in this, it appears random. But it is not *randomly generated*, it is random in that it has no meaning -> arbitrary.

Reply:
> it is not _randomly generated_

Of course, I did not say that this is the result of the generation of pre-random numbers

> it has no meaning

Not completely true, this data made sense until it was deallocated, and we can even try to read it and get some meaningful information.